### PR TITLE
Test Speedup HPO coverage

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds a functionality or fixes a bug.
OR
This PR marks a new Scout release. We apply semantic versioning. This is a major/minor/patch release for reasons.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
It gets stuck on the template page while printing stuff in this table:

```
<div class="table-responsive">
        <table class="table table-bordered table-hover">
        <thead>
          <tr>
            <th>{{ _('Sample') }}</th>
            <th>{{ _('Average coverage') }} [x]</th>
            {% for level_int, level_key in levels.items() %}
              <th>{{ _('Completeness') }} {{ level_int }}x [%]</th>
            {% endfor %}
          </tr>
        </thead>
        <tbody>
          {% for data in metrics_rows %}
            <tr>
              <td>{{ data.TranscriptStat.sample.name or data.TranscriptStat.sample.id }}</td>
              <td class="text-right">{{ data.mean_coverage|round(2) }}</td>
              {% for level_int, level_key in levels.items() %}
                <td class="text-right">{{ data|attr(level_key)|round(2) if data|attr(level_key) }}</td>
              {% endfor %}
            </tr>
          {% endfor %}
        </tbody>
      </table>
      </div>
```

**Comment by northwestwitch** [comment]
Yes, I think it's the query that creates the metrics that is slow, specifically this code [here](https://github.com/Clinical-Genomics/chanjo-report/blob/5660600f434c84f35f9142d412808202811aa1e4/chanjo_report/server/blueprints/report/utils.py#L96):

I'm going to move this issue to chanjo-report!

**Comment by northwestwitch** [comment]
This isn't fixed yet. I'm giving another try again. Would be nice to at least understand why the query gets so slow for the HPO panel of some cases!

**Comment by northwestwitch** [comment]
An update on this issue:

Looks like the slow query from my comment above might benefit by using an existing composite index that is not used, because if you explain the query the composite index `_sample_transcript_uc` is not used when filtering transcript_stat. Instead only the `where` condition is applied:
 
```
+----+-------------+-----------------+------------+-------+-------------------------------------+-----------------------+---------+-----------------------------+------+----------+-------------------------------------------+
| id | select_type | table           | partitions | type  | possible_keys                       | key                   | key_len | ref                         | rows | filtered | Extra                                     |
+----+-------------+-----------------+------------+-------+-------------------------------------+-----------------------+---------+-----------------------------+------+----------+-------------------------------------------+
|  1 | SIMPLE      | transcript      | NULL       | range | PRIMARY,ix_transcript_gene_id       | ix_transcript_gene_id | 4       | NULL                        |   40 |   100.00 | Using where; Using index; Using temporary |
|  1 | SIMPLE      | transcript_stat | NULL       | ref   | _sample_transcript_uc,transcript_id | transcript_id         | 34      | chanjo4-stage.transcript.id |  708 |     0.04 | Using where                               |
+----+-------------+-----------------+------------+-------+-------------------------------------+-----------------------+---------+-----------------------------+------+----------+-------------------------------------------+
```

**If you force the usage of the _sample_transcript_uc index** the query becomes much faster:

SELECT blabla bla from FROM transcript_stat force index (_sample_transcript_uc) bla bla bla:

0.28 secs VS 8.45 secs 

This is a good start. I just have to understand how to fix this in Alchy (which I don't know the syntax for)

File: scout/server/blueprints/cases/views.py [hunk 1]
```diff
@@ -49,6 +61,77 @@
...
+        api.query(
+            TranscriptStat,
+            func.avg(TranscriptStat.mean_coverage).label("mean_coverage"),
+            func.avg(TranscriptStat.completeness_10).label("completeness_10"),
+            func.avg(TranscriptStat.completeness_15).label("completeness_15"),
+            func.avg(TranscriptStat.completeness_20).label("completeness_20"),
+            func.avg(TranscriptStat.completeness_50).label("completeness_50"),
+            func.avg(TranscriptStat.completeness_100).label("completeness_100"),
+        )
+        .with_hint(TranscriptStat, "USE INDEX (_sample_transcript_uc)")
```
Comment (northwestwitch) [inline scout/server/blueprints/cases/views.py:76]: This is the fix that will be introduced in chanjo-report (see PR https://github.com/Clinical-Genomics/chanjo-report/pull/53)

Comment (dnil) [reply scout/server/blueprints/cases/views.py:76]: Well done! I think you can feel rather satisfied with that line of testing and problem solving! Simple fix in the end, but very much not obvious at first.

Comment (northwestwitch) [reply scout/server/blueprints/cases/views.py:76]: > Well done! I think you can feel rather satisfied with that line of testing and problem solving! Simple fix in the end, but very much not obvious at first.

No, I feel like I can become an SQL consultant from now on! 😆

**Comment by northwestwitch** [comment]
Closing, the issue has been fixed [in chanjo-report](https://github.com/Clinical-Genomics/chanjo-report/pull/53)

**Pull request closed without merging**